File: web/src/app/core/services/delegation.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { map } from 'rxjs/operators';
import {
  AgentCredit,
  Delegation,
  DeleguerAutoriteRequest,
  ReassignerDossierRequest,
} from '../models/delegation.model';

@Injectable({ providedIn: 'root' })
export class DelegationService {
  private readonly BASE = '/api/v1/delegations';
  private readonly CREDIT = '/api/v1/dossiers-credit';

  constructor(private http: HttpClient) {}

  listDelegations(page = 0, size = 20): Observable<{ content: Delegation[]; totalElements: number }> {
    return this.http
      .get<any>(this.BASE, { params: { page, size } })
      .pipe(map((r) => r.data));
  }

  mesDelegations(): Observable<Delegation[]> {
    return this.http
      .get<any>(`${this.BASE}/mes-delegations`)
      .pipe(map((r) => r.data));
  }

  getAgentsCredit(): Observable<AgentCredit[]> {
    return this.http
      .get<any>(`${this.BASE}/agents-credit`)
      .pipe(map((r) => r.data));
  }

  reassignerDossier(dossierUid: string, req: ReassignerDossierRequest): Observable<Delegation> {
    return this.http
      .patch<any>(`${this.CREDIT}/${dossierUid}/reassigner`, req)
      .pipe(map((r) => r.data));
  }

  deleguerAutorite(req: DeleguerAutoriteRequest): Observable<Delegation> {
    return this.http
      .post<any>(`${this.BASE}/deleguer-autorite`, req)
      .pipe(map((r) => r.data));
  }

  revoquer(uid: string): Observable<void> {
    return this.http
      .delete<any>(`${this.BASE}/${uid}/revoquer`)
      .pipe(map(() => void 0));
  }
}

File: mobile/lib/core/services/delegation_service.dart
import '../models/delegation.dart';
import '../models/page_response.dart';
import 'api_service.dart';

class DelegationService {
  final ApiService _api;

  DelegationService(this._api);

  Future<List<Delegation>> mesDelegations() async {
    return _api.get<List<Delegation>>(
      '/api/v1/delegations/mes-delegations',
      fromJson: (data) {
        final list = data as List<dynamic>;
        return list
            .map((e) => Delegation.fromJson(e as Map<String, dynamic>))
            .toList();
      },
    );
  }

  Future<PageResponse<Delegation>> listDelegations({
    int page = 0,
    int size = 20,
  }) async {
    return _api.get<PageResponse<Delegation>>(
      '/api/v1/delegations',
      queryParameters: {'page': page, 'size': size},
      fromJson: (data) => PageResponse.fromJson(
        data as Map<String, dynamic>,
        (json) => Delegation.fromJson(json),
      ),
    );
  }

  Future<List<AgentCreditItem>> getAgentsCredit() async {
    return _api.get<List<AgentCreditItem>>(
      '/api/v1/delegations/agents-credit',
      fromJson: (data) {
        final list = data as List<dynamic>;
        return list
            .map((e) => AgentCreditItem.fromJson(e as Map<String, dynamic>))
            .toList();
      },
    );
  }

  Future<Delegation> reassignerDossier({
    required String dossierUid,
    required String nouvelAgentUid,
    String? motif,
  }) async {
    return _api.patch<Delegation>(
      '/api/v1/dossiers-credit/$dossierUid/reassigner',
      data: {'nouvelAgentUid': nouvelAgentUid, if (motif != null) 'motif': motif},
      fromJson: (data) => Delegation.fromJson(data as Map<String, dynamic>),
    );
  }

  Future<void> revoquer(String delegationUid) async {
    await _api.delete('/api/v1/delegations/$delegationUid/revoquer');
  }
}
